keps/sig-node/0028-node-readiness++.md

```diff
@@ -0,0 +1,237 @@
+But the other newer system extensions may not be as mature yet, or may be slow to transition to “ready” for legitimate reasons.
+
+The lack of a single unambiguous signal on when a “node” is “ready”, beyond just the kubelet or a container runtime is causing several issues today.
+
+- System components like Cluster Autoscaler and Scheduler have to combine several Node Conditions to evaluate a node’s actual readiness. It is not straightforward to introduce new conditions into the policy.
+- Cluster Autoscaler incorrectly assumes a GPU node to be “ready” prior to device plugins and/or device drivers being installed which leads to spurious node additions and deletions thereby leading to custom logic in the autoscaler.
+- Logs & metrics are lost if logging & monitoring agents are not online which can lead to sub-optimal (often unexpected) user behavior.
+- A node with offline kube-proxy can result in service outage in the event of an application update (via deployments for example).
+- It is difficult for Node Controller and/or node repair systems to detect and remedy issues with extended system components of the node since they lack a set of generic & easy to identify signals from those individual components on the node.
+
+As Kubernetes becomes more and more customizable (or extensible) and gets deployed in mission critical systems, it is imperative to provide fail safe mechanisms that are built into the system.
+
+The remainder of this proposal explores a few approaches for addressing the issue of node health extensibility.
+
+## Solutions
+
+Any solution that is chosen for addressing the set of problems stated above should meet the following design requirements:
+
+- Be backwards compatible
+- Be extensible to accommodate variations in kubernetes environments - the set of system microservices can vary across k8s deployments.
+- Ideally, provide a single signal (not a policy) to cluster level components. 
+- Avoid introducing new API primitives (ideally).
+- Support self-bootstrapping - Scheduling system components (including the kubelet) even if the node is not logically ready.
+
+### Approach A - Track state of extensions via APIs (Preferred)
+
+For reasons similar to recent [node heartbeats revamp](https://github.com/kubernetes/community/blob/6a2652fdf1f7d662a70aa3202147b85ff2d6408e/keps/sig-node/0009-node-heartbeat.md), individual node extensions will be expected to post their state via dedicated "[Lease](https://github.com/kubernetes/api/blob/master/coordination/v1beta1/types.go#L27)" API objects.
+
+The Node API will include a new field that will track a list of node extensions that are expected on the node.
+This list can be initially populated by the node controller when a node object is created or by the kubelet as it’s extensions register, or by individual extensions themselves (if they have sufficient privileges). 
+
+Here are some sample illustrations:
+
+```yaml
+type Node struct {
+ …
+ Spec:
+  SystemExtensions []SystemExtension
```

Comment:
Is there a real need to add these to the Node spec? Couldn't they be tracked as separate objects each with a reference (i.e., name) to the node?

That would avoid any change to the actual node object here.